lib/block_view.go

```diff
@@ -4160,6 +4169,56 @@ func (bav *UtxoView) ConnectBlock(
 		}
 	}
 
+	// If we're past the PoS cutover, we need to track which validators were active.
+	if blockHeight >= uint64(bav.Params.ForkHeights.ProofOfStake2ConsensusCutoverBlockHeight) {
+		// Get the active validators for the block.
+		var signersList *bitset.Bitset
+		if !desoBlock.Header.ValidatorsVoteQC.isEmpty() {
+			signersList = desoBlock.Header.ValidatorsVoteQC.ValidatorsVoteAggregatedSignature.SignersList
+		} else {
+			signersList = desoBlock.Header.ValidatorsTimeoutAggregateQC.ValidatorsTimeoutAggregatedSignature.SignersList
+		}
+		allSnapshotValidators, err := bav.GetAllSnapshotValidatorSetEntriesByStake()
+		if err != nil {
+			return nil, errors.Wrapf(err, "ConnectBlock: error getting all snapshot validator set entries by stake")
+		}
+		currentEpochNumber, err := bav.GetCurrentEpochNumber()
+		if err != nil {
+			return nil, errors.Wrapf(err, "ConnectBlock: error getting current epoch number")
+		}
+		// TODO: Is it safe to assume that the leader is included in the signers list?
+		// If not, we should explicitly add them here since they were "active".
+		for ii, validator := range allSnapshotValidators {
+			// Skip validators who didn't sign
+			if !signersList.Get(ii) {
+				continue
+			}
+			// Get the current validator entry
+			validatorEntry, err := bav.GetValidatorByPKID(validator.ValidatorPKID)
+			if err != nil {
+				return nil, errors.Wrapf(err, "ConnectBlock: error getting validator by PKID")
+			}
+			// TODO: what do we want to do for jailed validators? I'm assuming we shouldn't hit them here.
+			if validatorEntry.Status() == ValidatorStatusJailed {
+				glog.Errorf("ConnectBlock: Validator %v is jailed but signed the block",
+					PkToString(validatorEntry.ValidatorPKID.ToBytes(), bav.Params))
+				continue
+			}
```

Comment:
```suggestion
			// It's possible for the validator to have unregistered since two epochs ago, but is continuing
			// to vote. If the validatorEntry is nil or IsDeleted, we skip it here.
			if validatorEntry == nil || validatorEntry.IsDeleted() {
				continue
			}

			// It's possible for the validator to be in the snapshot validator set, but to have been jailed
			// in the previous epoch due to inactivity. In the edge case where the validator now comes back
			// online, we maintain its jailed status until it unjails itself explicitly again.
			if validatorEntry.Status() == ValidatorStatusJailed {
				continue
			}
```